ql/src/java/org/apache/hadoop/hive/ql/exec/repl/DirCopyTask.java

```diff
@@ -203,4 +213,87 @@ public String getName() {
+              Collections.singletonList(sourcePath), targetPath, proxyUser,
+              conf, ShimLoader.getHadoopShims());
+      if (!result) {
+        LOG.error("Can not copy using snapshot diff for source: {} and "
+                + "target: {}. Falling back to normal copy.", sourcePath,
+            targetPath);
+        FileUtils.distCp(sourcePath.getFileSystem(conf), // source file system
+            Collections.singletonList(sourcePath), // list of source paths
+            targetPath, false, proxyUser, conf, ShimLoader.getHadoopShims());
+      }
+    } else if (getWork().getCopyMode()
+        .equals(SnapshotUtils.SnapshotCopyMode.INITIAL_COPY)) {
+      LOG.info("Using snapshot initial copy for source: {} and target: {}",
+          sourcePath, targetPath);
+      // Try allowSnapshot at target first, if success or already allowed
+      // will use snapshots to copy else fallback.
+      boolean isTargetSnapshottable =
+          SnapshotUtils.allowSnapshot(targetFs, targetPath);
+      if (!isTargetSnapshottable) {
+        // We can not allow creating snapshot at target, so no point moving
+        // to snapshot mode of copy, fallback to normal copy.
+        LOG.error("Can not copy from initial snapshot directory for source: {} "
+            + "and target: {}. Since target is not snapshottable. Falling "
+            + "back to normal copy.", sourcePath, targetPath);
+        FileUtils.distCp(sourcePath.getFileSystem(conf), // source file system
+            Collections.singletonList(sourcePath), // list of source paths
+            targetPath, false, proxyUser, conf, ShimLoader.getHadoopShims());
+        // Returning to avoid creating snapshots at target.
+        return;
+      } else {
+        // Get the path relative to the initial snapshot for copy.
+        Path snapRelPath = new Path(sourcePath,
+            HdfsConstants.DOT_SNAPSHOT_DIR + "/" + prefix + "initial");
+        boolean result = FileUtils.distCp(sourcePath.getFileSystem(conf), //
+            // source file system
+            Collections.singletonList(snapRelPath), // source path relative to
+            // snapshot
+            targetPath, false, proxyUser, conf, ShimLoader.getHadoopShims());
+        if (!result) {
+          LOG.error(
+              "Can not copy from initial snapshot directory for source: {} "
+                  + "and target: {}. Falling back to normal copy.", snapRelPath,
+              targetPath);
+          FileUtils.distCp(sourcePath.getFileSystem(conf), // source file system
+              Collections.singletonList(sourcePath), // list of source paths
+              targetPath, false, proxyUser, conf, ShimLoader.getHadoopShims());
+        }
+      }
+    }
+    // Create snapshot at target Filesystem, ignore exceptions if any,
+    // since copy is success, if creation fails, the next iteration shall
+    // handle.
+    SnapshotUtils.deleteSnapshotSafe(targetFs, targetPath, prefix + "old");
```

Comment:
this deletes older snapshot? Add a comment for clarity

Reply:
added explanation in the comment, Added javadoc to the util method as well, for posterity